bus-mapping/src/evm/opcodes/calldatacopy.rs

```diff
@@ -84,87 +79,126 @@ fn gen_calldatacopy_step(
     Ok(exec_step)
 }
 
-fn gen_memory_copy_step(
+fn gen_copy_steps(
     state: &mut CircuitInputStateRef,
     exec_step: &mut ExecStep,
     src_addr: u64,
     dst_addr: u64,
     src_addr_end: u64,
-    bytes_left: usize,
+    bytes_left: u64,
     is_root: bool,
-) -> Result<(), Error> {
-    for idx in 0..std::cmp::min(bytes_left, MAX_COPY_BYTES) {
-        let addr = src_addr + idx as u64;
-        let byte = if addr < src_addr_end {
+) -> Result<Vec<CopyStep>, Error> {
+    let mut copy_steps = Vec::with_capacity(2 * bytes_left as usize);
+    for idx in 0..bytes_left {
+        let addr = src_addr + idx;
+        let rwc = state.block_ctx.rwc;
+        let (value, is_pad) = if addr < src_addr_end {
             let byte =
                 state.call_ctx()?.call_data[(addr - state.call()?.call_data_offset) as usize];
             if !is_root {
                 state.push_op(
                     exec_step,
                     RW::READ,
-                    MemoryOp::new(state.call()?.caller_id, (addr as usize).into(), byte),
+                    MemoryOp::new(state.call()?.caller_id, addr.into(), byte),
```

Comment:
I used the `push_op` since the memory read is not from the current `call_id`, but from the call's `caller_id`. Since the `memory_read` assumes its the current call, I instead used a `push_op` specifying the `caller_id`.